### ISSUE TITLE
IOTA node panic with total order violation

### ISSUE THREAD

**Issue by junwei0117** [iotaledger/iota#4305]
## Bug description

We already encountered this twice in our testnet this week, causing our validators to panic and restart.

## Version

v0.7.2-rc

## Errors

We don’t have many logs for this error, here’s what I have at the moment.

```
2024-11-29 05:32:55.796	2024-11-28T21:32:55.795885Z  INFO iota_core::checkpoints::checkpoint_output: Creating checkpoint CheckpointDigest(6HKmeyBjAwqVBAbMGZULbnQ9AFJhFH2wvgmGNgfdBCWQ) at epoch 11, sequence 4047476, previous digest Some(CheckpointDigest(C6qH4AfYBi3oLmXmZd7eLSNW9Qm7AnnLeCutBczNMEtR)), transactions count 5, content digest CheckpointContentsDigest(BKCeBZMYA1XkbHM1CvGwHTfTTrfzrMj1PqJrBrieW7No), end_of_epoch_data None
2024-11-29 05:32:56.052	2024-11-28T21:32:56.052514Z  INFO iota_core::checkpoints::checkpoint_output: Creating checkpoint CheckpointDigest(FM4YToQd3bkbcrvqm4QRVs3ZAitXaJpX4cBzhofxBmj6) at epoch 11, sequence 4047477, previous digest Some(CheckpointDigest(6HKmeyBjAwqVBAbMGZULbnQ9AFJhFH2wvgmGNgfdBCWQ)), transactions count 4, content digest CheckpointContentsDigest(AkgN1Gf835d6EFLU11o331JksvxN1ndipoFn2iMMbkft), end_of_epoch_data None
2024-11-29 05:32:56.120	thread 'iota-node-runtime' panicked at core/src/slice/sort/shared/smallsort.rs:2024-11-28T21:32:56.120281Z ERROR telemetry_subscribers: panicked at core/src/slice/sort/shared/smallsort.rs:860:5:
2024-11-29 05:32:56.120	user-provided comparison function does not correctly implement a total order panic.file="core/src/slice/sort/shared/smallsort.rs" panic.line=860 panic.column=5
2024-11-29 05:32:56.120	860:5:
2024-11-29 05:32:56.120	user-provided comparison function does not correctly implement a total order
2024-11-29 05:32:56.120	stack backtrace:
2024-11-29 05:32:56.120	   0: rust_begin_unwind
2024-11-29 05:32:56.120	   1: core::panicking::panic_fmt
2024-11-29 05:32:56.120	   2: core::slice::sort::shared::smallsort::panic_on_ord_violation
2024-11-29 05:32:56.120	   3: core::slice::sort::shared::smallsort::small_sort_general_with_scratch
2024-11-29 05:32:56.120	   4: core::slice::sort::stable::quicksort::quicksort
2024-11-29 05:32:56.120	   5: core::slice::sort::stable::drift::sort
2024-11-29 05:32:56.120	   6: core::slice::sort::stable::driftsort_main
2024-11-29 05:32:56.120	   7: iota_network::state_sync::PeerBalancer::new
2024-11-29 05:32:56.120	   8: iota_network::state_sync::sync_one_checkpoint_contents::{{closure}}::{{closure}}
2024-11-29 05:32:56.120	   9: <futures_util::stream::futures_ordered::OrderWrapper<T> as core::future::future::Future>::poll
2024-11-29 05:32:56.120	  10: <futures_util::stream::futures_unordered::FuturesUnordered<Fut> as futures_core::stream::Stream>::poll_next
2024-11-29 05:32:56.120	  11: <futures_util::stream::futures_ordered::FuturesOrdered<Fut> as futures_core::stream::Stream>::poll_next
2024-11-29 05:32:56.120	  12: <tokio::future::poll_fn::PollFn<F> as core::future::future::Future>::poll
2024-11-29 05:32:56.120	  13: iota_network::state_sync::sync_checkpoint_contents::{{closure}}
2024-11-29 05:32:56.120	  14: tokio::runtime::task::core::Core<T,S>::poll
2024-11-29 05:32:56.120	  15: tokio::runtime::task::harness::Harness<T,S>::poll
2024-11-29 05:32:56.120	  16: tokio::runtime::scheduler::multi_thread::worker::Context::run_task
2024-11-29 05:32:56.120	  17: tokio::runtime::scheduler::multi_thread::worker::Context::run
2024-11-29 05:32:56.120	  18: tokio::runtime::context::scoped::Scoped<T>::set
2024-11-29 05:32:56.120	  19: tokio::runtime::context::runtime::enter_runtime
2024-11-29 05:32:56.120	  20: tokio::runtime::scheduler::multi_thread::worker::run
2024-11-29 05:32:56.120	  21: tokio::runtime::task::core::Core<T,S>::poll
2024-11-29 05:32:56.120	  22: tokio::runtime::task::harness::Harness<T,S>::poll
2024-11-29 05:32:56.120	  23: tokio::runtime::blocking::pool::Inner::run
2024-11-29 05:32:56.120	note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
2024-11-29 05:33:06.572	2024-11-28T21:33:06.572378Z  INFO iota_node: Iota Node version: 0.7.2-rc-bf98615df726
2024-11-29 05:33:06.572	2024-11-28T21:33:06.572389Z  INFO iota_node: Supported protocol versions: Some(SupportedProtocolVersions { min: ProtocolVersion(1), max: ProtocolVersion(1) })
2024-11-29 05:33:06.572	2024-11-28T21:33:06.572395Z  INFO iota_node: Started Prometheus HTTP endpoint at 0.0.0.0:9184
```

**Comment by thibault-martinez** [comment]
This should be the fix https://github.com/iotaledger/iota/pull/4211.
0.7.3 including it is coming soon I think.

**Comment by alexsporn** [comment]
Should be fixed with 0.7.3-rc.
Please open a new issue if this happens again.
